src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php

```diff
@@ -402,4 +406,38 @@ protected function addLink(Request $request, LinkInterface $link): void
 
         $request->attributes->set('_links', $linkProvider->withLink($link));
     }
+
+    /**
+     * @param LinkInterface[] $links
+     */
+    protected function sendEarlyHints(iterable $links, Response $response = null): Response
+    {
+        if (!class_exists(HttpHeaderSerializer::class)) {
+            throw new \LogicException('You cannot use the "sendEarlyHints" method if the WebLink component is not available. Try running "composer require symfony/web-link".');
+        }
+
+        if (null === $response) {
+            $response = new Response();
+        }
+
+        if (null === $this->httpHeaderSerializer) {
+            $this->httpHeaderSerializer = new HttpHeaderSerializer();
```

Comment:
let's get this via the subscriber interface, that'd give some more flexibility

Reply:
This class has no associated interface (on purpose), but I switched to a service subscription.